django/template/loaders/filesystem.py

```diff
@@ -20,8 +20,15 @@ def get_dirs(self):
 
     def get_contents(self, origin):
         try:
-            with open(origin.name, encoding=self.engine.file_charset) as fp:
-                return fp.read()
+            # Read and decode file contents with fewer system calls than plain fp.read().
+            with open(origin.name, 'rb', buffering=0) as fp:
+                contents = bytearray()
+                while True:
+                    data = fp.read(65536)
+                    if not data:
+                        break  # end of file
+                    contents += data
+                return contents.decode(self.engine.file_charset)
```

Comment:
This could be a little more compact:

```suggestion
            contents = bytearray()
            with open(origin.name, 'rb', buffering=0) as fp:
                while data := fp.read(65536):
                    contents += data
            return contents.decode(self.engine.file_charset)
```